src/main/java/edu/ucsb/mapache/controllers/RoleController.java

```diff
@@ -100,4 +108,41 @@
     String body = mapper.writeValueAsString(response);
     return ResponseEntity.ok().body(body);
   }
-}
+
+  // Added two functions below to modify DB to add API Token field
+  
+  @GetMapping("/apiKey")
+  public ResponseEntity<String> getCustomAPIToken(@RequestHeader("Authorization") String authorization)
+      throws JsonProcessingException {
+        AppUser user = authControllerAdvice.getUser(authorization);
+        String userToken = user.getApiToken();
+        Map<String, String> response = new HashMap<>();
+        response.put("token", userToken);
+        String body = mapper.writeValueAsString(response);
+        return ResponseEntity.ok().body(body);
+        
+  }
+
+  //Sets API Token
+  @PutMapping(value = "/addApiKey")
+  public ResponseEntity<String> setCustomApiToken(@RequestHeader("Authorization") String authorization, 
+  @RequestBody @Valid String token) 
+      throws JsonProcessingException {
+        AppUser user = authControllerAdvice.getUser(authorization);
+        SearchParameters sp = new SearchParameters();
+        sp.setQuery("empty");
+        sp.setPage(1);
+        
+        String i = googleSearchService.getJSON(sp, token);
```

Comment:
I personally don't really like the idea of burning through one search query just to test if a token is valid before setting it. Is it really necessary to check if the token is valid here? I think this should be a problem for the actual search itself to handle, and not for the setter.

Reply:
I see. Alternatively, whenever a search is performed, we could clear the api token if there is an error. But the problem is that there could be other things causing the search to return an error, so checking then might not be catching the correct error (and could cause you to reset your token when it is a valid token). Haoming and I searched for another way to check validity for custom search particularly but were unsuccessful.

Reply:
In any case, we probably shouldn't be clearing the key if there's an error. That's up to the user using the app to do, not us. Instead, we should just warn the user with an error message stating an error happened, and if the error is one of an invalid key, we can instruct them to visit this page and change / clear their token.

Reply:
Two things:
* I am ok with burning one search to check the validity of the token.   I think that should stay in the code base.   You get 100 per day, and I'd rather find out immediately than have to go do a search and get an error message in  a place where I can't fix the problem.
* On the second point, I think Andrew's correct; the UX here isn't ideal. But let's not make the perfect the enemy of the good.  Let's get this into the code base, and then adjust the UX.  Delaying getting this because it isn't perfect in serves no one's interests.

Reply:
I just thought of another possible (though highly unlikely) edge case - if a user inputs a token that has used up its daily quota but is still valid, this check won't allow it. My initial thought for removing this was that it has the potential to create some really strange problems that may not be easily debugged, so to save some weird confusion later, I figured we should just let the search page do all the checking / error handling instead. 

But I agree with Conrad that these issues can be addressed later and that we should probably get this merged in (especially given that the quarter is now over).